sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala

```diff
@@ -236,4 +237,24 @@ class WholeStageCodegenSuite extends QueryTest with SharedSQLContext {
       }
     }
   }
+
+  test("SPARK-22551: Fix 64kb limit for deeply nested expressions under wholestage codegen") {
+    import testImplicits._
+    withTempPath { dir =>
+      val path = dir.getCanonicalPath
+      val df = Seq(("abc", 1)).toDF("key", "int")
+      df.write.parquet(path)
+
+      var strExpr: Expression = col("key").expr
+      for (_ <- 1 to 150) {
+        strExpr = Decode(Encode(strExpr, Literal("utf-8")), Literal("utf-8"))
+      }
+      val expressions = Seq(If(EqualTo(strExpr, strExpr), strExpr, strExpr))
+
+      val df2 = spark.read.parquet(path).select(expressions.map(Column(_)): _*)
+      val plan = df2.queryExecution.executedPlan
+      assert(plan.find(_.isInstanceOf[WholeStageCodegenExec]).isDefined)
```

Comment:
can you give some insights about how this test fail without your PR? In `WholeStageCodegen.doExec`, we have
```
  val (ctx, cleanedSource) = doCodeGen()
    // try to compile and fallback if it failed
    val (_, maxCodeSize) = try {
      CodeGenerator.compile(cleanedSource)
    } catch {
      case _: Exception if !Utils.isTesting && sqlContext.conf.codegenFallback =>
        // We should already saw the error message
        logWarning(s"Whole-stage codegen disabled for this plan:\n $treeString")
        return child.execute()
    }
```

Reply:
The error message is:
```
[info]   Cause: org.codehaus.janino.InternalCompilerException: Code of method "processNext()V" of class "org.apache
.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator" grows beyond 64 KB
[info]   at org.codehaus.janino.CodeContext.makeSpace(CodeContext.java:990)
[info]   at org.codehaus.janino.CodeContext.write(CodeContext.java:867)
[info]   at org.codehaus.janino.UnitCompiler.writeOpcode(UnitCompiler.java:11901)
[info]   at org.codehaus.janino.UnitCompiler.pushConstant(UnitCompiler.java:10362)
[info]   at org.codehaus.janino.UnitCompiler.compileGet2(UnitCompiler.java:5231)
[info]   at org.codehaus.janino.UnitCompiler.access$8500(UnitCompiler.java:212)
[info]   at org.codehaus.janino.UnitCompiler$12.visitIntegerLiteral(UnitCompiler.java:4099)
[info]   at org.codehaus.janino.UnitCompiler$12.visitIntegerLiteral(UnitCompiler.java:4070)
[info]   at org.codehaus.janino.Java$IntegerLiteral.accept(Java.java:5282)
[info]   at org.codehaus.janino.UnitCompiler.compileGet(UnitCompiler.java:4070)
[info]   at org.codehaus.janino.UnitCompiler.fakeCompile(UnitCompiler.java:3406)
[info]   at org.codehaus.janino.UnitCompiler.compileGetValue(UnitCompiler.java:5247)
[info]   at org.codehaus.janino.UnitCompiler.compileContext2(UnitCompiler.java:4033)
[info]   at org.codehaus.janino.UnitCompiler.access$6000(UnitCompiler.java:212)
```

Reply:
Since it is in testing (`Utils.isTesting`), the compilation error won't be ignored.